concrete/src/Updater/Migrations/Migrations/Version20181116072400.php

```diff
@@ -50,35 +51,25 @@ protected function fixPublishStartEndDates()
      */
     protected function fixOverlappingStartEndDates()
     {
-        $rs = $this->connection->executeQuery(<<<'EOT'
-select
-    cv1.cID,
-    cv1.cvID
-from
-    CollectionVersions as cv1
-    inner join CollectionVersions as cv2
-        on cv1.cID = cv2.cID
-        and cv1.cvID <> cv2.cvID
-        and cv1.cvIsApproved = cv2.cvIsApproved
-where
-    cv1.cvIsApproved = 1
-    and ifnull(cv1.cvPublishDate, '0000-00-00 00:00:00') <= ifnull(cv1.cvPublishEndDate, '9999-99-99 99:99:99')
-    and ifnull(cv1.cvPublishEndDate, '9999-99-99 99:99:99') >= ifnull(cv2.cvPublishDate, '0000-00-00 00:00:00')
-order by
-    cv1.cID,
-    cv1.cvID desc
-EOT
-        );
-        $page = null;
-        while (($row = $rs->fetch(PDO::FETCH_ASSOC)) !== false) {
-            if ($page === null || $page->getCollectionID() != $row['cID']) {
-                $page = Page::getByID($row['cID'], $row['cvID']);
-            } else {
-                $page->loadVersionObject($row['cvID']);
-            }
+        $qb = $this->connection->createQueryBuilder();
+        $result = $qb->select('p.cID')
+            ->from('Pages', 'p')
+            ->innerJoin('p', 'Collections', 'c', 'p.cID = c.cID')
+            ->innerJoin('p', 'CollectionVersions', 'cv', 'p.cID = cv.cID')
+            ->andWhere('p.cPointerID < 1')
+            ->andWhere('p.cIsTemplate = 0')
+            ->andWhere('cv.cvID = (select cvID from CollectionVersions where cID = cv.cID and cvIsApproved = 1 and (cvPublishDate is not null or cvPublishEndDate is not null) order by cvPublishDate desc limit 1)')
+            ->execute()
+        ;
+
+        while ($row = $result->fetch()) {
+            $page = Page::getByID($row['cID'], 'SCHEDULED');
```

Comment:
What if there are more versions that needs to be fixed? Here we only fix one version (the next "scheduled" one), but we may have more versions with overlapping dates...

Reply:
@mlocati That's the point I'm talking about. `setPublishInterval` function calls `avoidApprovalOverlapping` function, and `avoidApprovalOverlapping` function will fix publish dates for all other versions. That means we can fix all versions with overlapping dates by just calling `setPublishInterval` one time for recent version.

See: 
 https://github.com/concrete5/concrete5/blob/da7a3370f50e893093bb134029bceb11221e489f/concrete/src/Page/Collection/Version/Version.php#L943-L1048

Reply:
And also, if we call `avoidApprovalOverlapping` for a page, it breaks page scheduling.